### PR TITLE
Add @addons endpoint

### PR DESCRIPTION
**Pull request by esteele**
Beginning to add the `@addons` endpoint as per our discussion earlier today.

`GET /@addons` : Lists all addons
`GET /@addons/package.name` : Shows details of a particular addon
`POST /@addons/package.name/install`: Installs the addon
`POST /@addons/package.name/uninstall`: Uninstalls the addon
`POST /@addons/package.name/upgrade`: Upgrades addon to latest version

### CODE REVIEW

File: docs/source/_json/addons_get.resp [hunk 1]
```diff
@@ -0,0 +1,36 @@
+{
+  "description": "Optional plone.session refresh support.",
+  "profile_type": "default",
+  "upgrade_profiles": {},
+  "uninstall_profile_id": "plone.session:uninstall",
+  "other_profiles": [],
+  "is_installed": "FALSE",
```
Comment (lukasgraf) [inline docs/source/_json/addons_get.resp]: That should probably be serialized to JavsScript `false`.

File: docs/source/_json/addons_get.resp [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  "id": "plone.session",
+  "install_profile": {
+    "product": "plone.session",
+    "description": "Optional plone.session refresh support.",
+    "for": "NULL",
+    "title": "Session refresh support",
+    "pre_handler": "NULL",
+    "version": "1001",
+    "path": "/applicationPath/api/eggs/plone.session-3.7.0-py2.7.egg/plone/session/profiles/default",
```
Comment (lukasgraf) [inline docs/source/_json/addons_get.resp]: Do we really want to expose (absolute) filesystem paths? (Event though the endpoint is protected with `ManagePortal`)

File: docs/source/_json/addons_get_list.resp [hunk 3]
```diff
@@ -1,3 +1,51 @@
...
+  "@id": "http://localhost:55001/plone/@addons",
+  "batching": {
+    "@id": "http://localhost:55001/plone/@addons",
+    "first": "http://localhost:55001/plone/@addons?b_start=0",
+    "last": "http://localhost:55001/plone/@addons?b_start=1500",
+    "next": "http://localhost:55001/plone/@addons?b_start=25"
+  },
+  "items": [
+    {
+      "description": "Optional plone.session refresh support.",
```
Comment (lukasgraf) [inline docs/source/_json/addons_get_list.resp]: Each item should also get an `@id` property with the URL for the individual addon record (same for the response for individual addons).
E.g. `"@id": "http://localhost:55001/plone/@addon/plone.session"`

File: docs/source/_json/addons_install.resp [hunk 4]
```diff
@@ -0,0 +1,6 @@
+HTTP/1.1 200 OK
```
Comment (lukasgraf) [inline docs/source/_json/addons_install.resp]: We might want to mirror the behavior of [`POST` for content CRUD operations](http://plonerestapi.readthedocs.io/en/latest/content.html#successful-response-204-no-content) here:
- Return `204 No Content` to just acknowledge success (we don't really have anything else to say here)
- Return `200 OK` with the new representation of the "addon resource" if the client requests it via the `Prefer: return=representation` header

I think the same should be applicable to `uninstall` and `upgrade`.

File: docs/source/index.rst [hunk 5]
```diff
@@ -22,6 +22,7 @@ Contents
    content
    history
    batching
+   addons
```
Comment (lukasgraf) [inline docs/source/index.rst]: FYI: If you want to supress the Sphinx warning, but don't want it included in the toctree yet, you can add `:orphan:` at the top of the file.

**Comment by tisto** [comment]
FYI: Did a quick review of the PR. Overall this looks fine to me, lots of static code analysis violations and test failures though. Needs a bit more work it seems.

**Comment by tisto** [comment]
FYI: I removed the 4.0.0 milestone. We have 4 test failures and it does not seem this enhancement will make it.

**Comment by rodfersou** [comment]
@esteele I just merge master here and run tests locally (which pass), and also did some tests with Volto.
Everything looks fine here, is it okay for you to merge?

**Comment by rodfersou** [comment]
Well, the uninstall addon test is not working just on travis, and for what I look it should works..

@sneridagh @csenger is it possible for you help me to figure out why it fails just on travis?

**Comment by tisto** [comment]
@rodfersou the failing tests have a "not implemented yet" comment:

https://github.com/plone/plone.restapi/blob/addons-endpoint/src/plone/restapi/tests/test_addons.py#L110

We either have to write those tests of comment them out.

**Comment by tisto** [comment]
No way to get back to the control panel overview:

Addons

@rodfersou could you imagine to look into this?

**Comment by rodfersou** [comment]
@jenkins-plone-org please run jobs

**Comment by rodfersou** [comment]
@jenkins-plone-org please run jobs

**Comment by rodfersou** [comment]
@jenkins-plone-org please run jobs

**Comment by rodfersou** [comment]
@tisto finished the tests; run black; fix code analysis

but still not working for Plone 4.3... need more deep understand on what is missing to work.

for example, the uninstall test in Plone 4.3 is returning 204, but the addon status remain installed

other Plones are green

**Comment by rodfersou** [comment]
@jenkins-plone-org please run jobs

**Comment by rodfersou** [comment]
finally green at Travis \o/

File: news/733.feature [hunk 6]
```diff
@@ -0,0 +1 @@
+Add endpoints for managing addons. [esteele]
```
Comment (tisto) [inline news/733.feature:1]: @rodfersou pls add your name here as well for credit. We can do this after the merge to not have to trigger another Jenkins job.

Comment (rodfersou) [reply news/733.feature:1]: will do

Comment (esteele) [reply news/733.feature:1]: Fantastic work @rodfersou. Thank you for making this happen!

Comment (esteele) [reply news/733.feature:1]: Fantastic work @rodfersou. Thank you for making this happen!

Comment (rodfersou) [reply news/733.feature:1]: I just did final touches, the fantastic work is your @esteele 😉

**Comment by rodfersou** [comment]
@jenkins-plone-org please run jobs

**Comment by sneridagh** [comment]
@tisto this one is ready, the frontend as well. 😊

**Pull request merged**